src/callgraphgenerator.h

```diff
@@ -0,0 +1,74 @@
+/*
+  callgraphgenerator.h
+
+  This file is part of Hotspot, the Qt GUI for performance analysis.
+
+  Copyright (C) 2021 Klarälvdalens Datakonsult AB, a KDAB Group company, info@kdab.com
+  Author: Lieven Hey <lieven.hey@kdab.com>
+
+  Licensees holding valid commercial KDAB Hotspot licenses may use this file in
+  accordance with Hotspot Commercial License Agreement provided with the Software.
+
+  Contact info@kdab.com if any conditions of this licensing are not clear to you.
+
+  This program is free software; you can redistribute it and/or modify
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation, either version 2 of the License, or
+  (at your option) any later version.
+
+  This program is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#pragma once
+
+#include "data.h"
+#include "models/callercalleemodel.h"
+
+class QTextStream;
+class QModelIndex;
+
+template<typename Map>
+void iterateModel(int height, int role, const Data::Symbol& symbol, CallerCalleeModel* model, uint parent,
+                  QTextStream& stream, QHash<QString, QModelIndex>& lookup, float threshold)
+{
+    if (height == 0) {
+        return;
+    }
+
+    if (symbol.prettySymbol.isEmpty())
+        return;
+
+    const auto index = model->indexForSymbol(symbol);
+    const auto map = index.data(role).value<Map>();
+    const auto node = QStringLiteral("node%1 [label=\"%2\"]\n");
+    const auto connect = role == CallerCalleeModel::CallersRole ? QStringLiteral("node%2 -> node%1\n")
+                                                                : QStringLiteral("node%1 -> node%2\n");
+
+    const auto costIndex = model->index(index.row(), CallerCalleeModel::NUM_BASE_COLUMNS);
+    const auto total = costIndex.data(CallerCalleeModel::TotalCostRole).toLongLong();
+
+    for (const auto& key : map.keys()) {
+        const uint nodeId = qHash(key, role);
+        const auto nodeString = QStringLiteral("node%1").arg(nodeId);
+
+        if (lookup.contains(nodeString))
+            continue;
+
+        const auto cost = map[key][0];
+        if (static_cast<double>(cost) / total * 100 < threshold) {
```

Comment:
so the threshold is a percentage? that should be made clear in the name, something like `thresholdPercentage` - and ideally you'd make that one be in the range [0..1] and then remove the `* 100` multiplication